<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>plot2up by phraber</title>

    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/pygment_trac.css">
    <script src="javascripts/scale.fix.js"></script>
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="wrapper">
      <header>
        <h1 class="header">plot2up</h1>
        <p class="header">Generates a PDF paired tree &amp; pixel plot for each fasta file in cwd.</p>

        <ul>
          <li class="download"><a class="buttons" href="https://github.com/phraber/plot2up/zipball/master">Download ZIP</a></li>
          <li class="download"><a class="buttons" href="https://github.com/phraber/plot2up/tarball/master">Download TAR</a></li>
          <li><a class="buttons github" href="https://github.com/phraber/plot2up">View On GitHub</a></li>
        </ul>

        <p class="header">This project is maintained by <a class="header name" href="https://github.com/phraber">phraber</a></p>


      </header>
      <section>
        <h1>
<a id="plot2up" class="anchor" href="#plot2up" aria-hidden="true"><span class="octicon octicon-link"></span></a>plot2up</h1>

<p>Generates a PDF paired tree &amp; pixel plot for each fasta file in cwd.
plot2up, version 0.0.1a, 23 August 2013
PURPOSE: Generates a PDF paired tree &amp; pixel plot for each fasta file in cwd.
  USAGE: type 'make'
AUTHORS: Peter Hraber, Will Fischer, Bette Korber</p>

<p>I. DEPENDENCIES</p>

<p>Currently assuming Mac OS X environment, with interactive console.</p>

<p>A. Executables</p>

<p>Required:
  perl
        The perl interpreter (or a link to it) should be in /usr/bin.</p>

<p>make or gmake
        Apple/OS X Developer command-line tools. 
        See <a href="http://www.macports.org/install.php">http://www.macports.org/install.php</a>
  R     <a href="http://cran.r-project.org/bin/macosx">http://cran.r-project.org/bin/macosx</a>
        To ensure integrity of the dowloaded file, run 'md5' command on
        it and compare with published value.</p>

<p>Optional:
  phyml
       Not strictly necessary but helpful for all but the largest alignments.</p>

<pre><code>    If you choose not to use phyml to make trees, you will have to
    customize the Makefile and/or ensure the Newick tree file
    contains the same prefix as the Fasta alignment file.  Be
    aware that make might over-write this file.
</code></pre>

<p>Newick utilities
       Currently used only for nw_reroot but has great general utility.
       <a href="http://cegg.unige.ch/newick_utils">http://cegg.unige.ch/newick_utils</a></p>

<p>B. C library</p>

<p>libgd
        This is the trickiest component to build and install from source.
        Using macports to obtain the pre-built 'gs' will install this for you.
        Install the library before installing the perl module below.
        See <a href="http://www.macports.org/install.php">http://www.macports.org/install.php</a></p>

<p>C. PERL module/s</p>

<p>bioperl
        See <a href="http://www.bioperl.org/wiki/Installing_BioPerl_on_Unix">http://www.bioperl.org/wiki/Installing_BioPerl_on_Unix</a>
  GD
        perl -MCPAN -e 'install "GD"'  (Easy, right?)</p>

<p>D. R packages</p>

<p>png
    <a href="http://cran.r-project.org/package=png">http://cran.r-project.org/package=png</a></p>

<p>ape
    <a href="http://cran.r-project.org/package=ape">http://cran.r-project.org/package=ape</a></p>

<p>To install these from within R using install.packages(), firewall rules 
  require that the proxy server be set as follows:
      export HTTP_PROXY=<a href="http://proxyout.lanl.gov:8080">http://proxyout.lanl.gov:8080</a>
      export HTTPS_PROXY=<a href="https://proxyout.lanl.gov:8080">https://proxyout.lanl.gov:8080</a>
      or within R, Sys.setenv("HTTP_PROXY=<a href="http://proxyout.lanl.gov:8080">http://proxyout.lanl.gov:8080</a>"), etc.</p>

<p>Otherwise, manually download and install them from the command line, e.g:
    R CMD INSTALL ape_3.0-9.tgz
  You will have to download and install the other required packages:
    <a href="http://cran.r-project.org/package=nlme">http://cran.r-project.org/package=nlme</a>
    <a href="http://cran.r-project.org/package=gee">http://cran.r-project.org/package=gee</a>
    <a href="http://cran.r-project.org/package=expm">http://cran.r-project.org/package=expm</a></p>

<p>II. Known Issues</p>

<ul>
<li>Files are mistreated when inserted into MS Office ppt or doc files.</li>
<li>Goal is to replace png files generated by pixel with pure-R raster images.</li>
<li>This version assumes nucleotide sequences.</li>
</ul>

<p>FIN</p>
      </section>
      <footer>
        <p><small>Hosted on <a href="http://pages.github.com">GitHub Pages</a> using the Dinky theme</small></p>
      </footer>
    </div>
    <!--[if !IE]><script>fixScale(document);</script><![endif]-->
		
  </body>
</html>
